### ISSUE TITLE
Setting icons?

### ISSUE THREAD

**Issue by dnish** [raix/push#132]
Hey,
I just want to ask, if it possible, to set some custom icons for a notification? In my Android app, I have an colored icon in my notification bar, but I want to set the "gray" one like other apps also do.

**Comment by jpgilchrist** [comment]
@dnish you need to setup your icons in your mobile-config.js file by establishing your

```javascript
App.icons({
  // iOS
  'iphone': 'resources/icons/icon-60x60.png',
  'iphone_2x': 'resources/icons/icon-60x60@2x.png',
  'iphone_3x': 'resources/icons/icon-60x60@3x.png',
  'ipad': 'resources/icons/icon-76x76.png',
  'ipad_2x': 'resources/icons/icon-76x76@2x.png',

  // Android
  'android_ldpi': 'resources/icons/icon-36x36.png',
  'android_mdpi': 'resources/icons/icon-48x48.png',
  'android_hdpi': 'resources/icons/icon-72x72.png',
  'android_xhdpi': 'resources/icons/icon-96x96.png'
});
```

**Comment by pmwisdom** [comment]
So I can't figure out how to do this. But in the notification bar, for example, those resolutions work correctly. But when you open up your "action bar" (swipe from the top down) the notificaiton icon appears too small. Which corresponds to different sized icons as seen here:

Action bar, Dialog & Tab icons
24 × 24 area in 32 × 32 (mdpi)
36 × 36 area in 48 × 48 (hdpi)
48 × 48 area in 64 × 64 (xhdpi)
72 × 72 area in 96 × 96 (xxhdpi)
96 × 96 area in 128 × 128 (xxxhdpi)

vs 

Notification icons
22 × 22 area in 24 × 24 (mdpi)
33 × 33 area in 36 × 36 (hdpi)
44 × 44 area in 48 × 48 (xhdpi)
66 × 66 area in 72 × 72 (xxhdpi)
88 × 88 area in 96 × 96 (xxxhdpi)

http://iconhandbook.co.uk/reference/chart/android/

Or maybe i'm missing something.

**Comment by dnish** [comment]
@pmwisdom I think we have to set both icons manually. I've used another Cordova plugin, and there I've to set "smallIcon" and "icon" for both (smallicon for notification bar and the normal icon for the action bar).

**Comment by pmwisdom** [comment]
@dnish  Yes you are correct. I had to do some native android work a bit ago and to get it to work I had to both set notification.setSmallIcon(icon) and notification.setLargeIcon(icon).

In the push plugin docs it says :

"Additionally, each push can include a large icon which is used to personalize each push."

**Comment by williamli** [comment]
So am I correct to say that if I do what to have a special grey icon to show when Android receive a push alert, I will need to build the PhoneGap project and then try to add it from Android Studio?

@pmwisdom where did you read that you can put a large icon with each push? I don't see it anywhere in the raix/push doc

**Comment by pmwisdom** [comment]
@williamli I scoured the code and I don't believe Raix has added support in his interface for adding large images. The actual phonegap plugin does support it though. 

Check out : https://github.com/phonegap/phonegap-plugin-push#images. That's the plugin that's used in this package.

**Comment by williamli** [comment]
@pmwisdom any idea how I can hook into with a new drawable icon?

var push = PushNotification.init({ 
        "android": { 
            "senderID": "123456789", "icon": "phonegap", "iconColor": "blue"}, 
        "ios": {"alert": "true", "badge": "true", "sound": "true"}, "windows": {} 
    });

**Comment by williamli** [comment]
@pmwisdom just scan through part of the code again. it seems that the android icon and colour is mentioned in the code, just not documented: https://github.com/raix/push/blob/f08079304c1b69897ed05edc7d96954d0416e35e/plugin/push.configuration.js#L100-L106

**Comment by williamli** [comment]
yay! I got the icon colour working by setting "iconColor": "#FF1744" in the config.push.json file.
still trying my luck with the icon though.

**Comment by williamli** [comment]
no idea how to make a drawable out of a new icon. but iconColor can be defined. If someone figured out how to make a drawable, you can set it as part of the config.push.json like https://github.com/raix/push/blob/f08079304c1b69897ed05edc7d96954d0416e35e/plugin/push.configuration.js#L100-L106

**Comment by povesteam** [comment]
@williamli cc @raix 

A custom push icon (for Android at least) can be set this way:
Create the .png icon in `your_project/resources/icons/pushicon.png` (or any other path you like)
Create an executable file `your_project/cordova-build-override/hooks/after_prepare/copy_pushicon.sh` (any file name, but keep the dir names exact) and write something similar inside:

```
#!/bin/bash

set -e #exit on any error

PROJECT_PATH=~/your_project
LOGFILE=~/pushicon_copy.log
DRAWABLE_DIR=$PROJECT_PATH/.meteor/local/cordova-build/platforms/android/res/drawable

mkdir -p $DRAWABLE_DIR >>$LOGFILE 2>&1
cp $PROJECT_PATH/resources/icons/pushicon.png $DRAWABLE_DIR/pushicon.png >>$LOGFILE 2>&1
```

Credit: http://devgirl.org/2013/11/12/three-hooks-your-cordovaphonegap-project-needs/

**Comment by raix** [comment]
@povesteam couldn't you just add the pushicon.png in the `cordova-build-override/platforms/android/res/drawable/pushicon.png` (instead of the prepare hook?)

**Comment by povesteam** [comment]
@raix I tried that first. :smile: 
But that way `meteor build` crashes when Cordova compiles for Android, telling me that the `android` dir isn't an Android application.

**Comment by raix** [comment]
did you report that as an issue to mdg :)

**Comment by povesteam** [comment]
Not yet, but I will.
Thanks for the great push package, btw. :+1:

**Comment by lampe** [comment]
hey,

i have created a file in my-project/cordova-build-override/platforms/android/res/drawable/pushicon.png
and added that to my config.push.json:
```javascript
{
  ....
  "icon": "pushicon",
  "iconColor": "#FF1744"
 ....
}
```
when i look into the build folder i can see that the pushicon.png file is there but i still get a white square in my notification bar on android.
when i pull the notification down i can see the iconColor but i see the square again there too.
Can someone help ?

**Comment by williamli** [comment]
Yes, your icon need to be a png file with alpha (transparency). Android will colour in your icon in white with the background being transparent.

**Comment by VioletaCalvo** [comment]
Hi!

Adding icons in `cordova-build-override/platforms/android/res/drawable/pushicon.png` worked perfect for me, thank you!
To generate all icons uder `cordova-build-override/platforms/android/res/`I used [Android Asset Studio - Notification Icon Generator](https://romannurik.github.io/AndroidAssetStudio/icons-notification.html#source.type=text&source.space.trim=1&source.space.pad=0&source.text.text=b&source.text.font=Pacifico&name=ic_stat_b). Hope this help someone :blush:

**Comment by adam-hanna** [comment]
@pmwisdom did you ever fix the small icon problem? I don't want custom icons, I just want my normal icon to show up in my push notification lol haha

**Comment by sahanDissanayake** [comment]
Get a huge error after @VioletaCalvo instructions on last comment

```
ANDROID_HOME=/Users/sahandissanayake/Library/Android/sdk
JAVA_HOME=/Library/Java/JavaVirtualMachines/jdk1.7.0_79.jdk/Contents/Home
:cdvCreateAssetManifest
:preBuild
:preReleaseBuild
:checkReleaseManifest
:CordovaLib:preBuild UP-TO-DATE
:CordovaLib:preReleaseBuild UP-TO-DATE
:CordovaLib:compileReleaseNdk UP-TO-DATE
:CordovaLib:compileLint
:CordovaLib:copyReleaseLint UP-TO-DATE
:CordovaLib:mergeReleaseProguardFiles UP-TO-DATE
:CordovaLib:packageReleaseRenderscript UP-TO-DATE
:CordovaLib:checkReleaseManifest
:CordovaLib:prepareReleaseDependencies
:CordovaLib:compileReleaseRenderscript UP-TO-DATE
:CordovaLib:generateReleaseResValues UP-TO-DATE
:CordovaLib:generateReleaseResources UP-TO-DATE
:CordovaLib:packageReleaseResources UP-TO-DATE
:CordovaLib:compileReleaseAidl UP-TO-DATE
:CordovaLib:generateReleaseBuildConfig UP-TO-DATE
:CordovaLib:generateReleaseAssets UP-TO-DATE
:CordovaLib:mergeReleaseAssets UP-TO-DATE
:CordovaLib:processReleaseManifest UP-TO-DATE
:CordovaLib:processReleaseResources UP-TO-DATE
:CordovaLib:generateReleaseSources UP-TO-DATE
:CordovaLib:compileReleaseJavaWithJavac UP-TO-DATE
:CordovaLib:processReleaseJavaRes UP-TO-DATE
:CordovaLib:transformResourcesWithMergeJavaResForRelease UP-TO-DATE
:CordovaLib:transformClassesAndResourcesWithSyncLibJarsForRelease UP-TO-DATE
:CordovaLib:mergeReleaseJniLibFolders UP-TO-DATE
:CordovaLib:transformNative_libsWithMergeJniLibsForRelease UP-TO-DATE
:CordovaLib:transformNative_libsWithSyncJniLibsForRelease UP-TO-DATE
:CordovaLib:bundleRelease UP-TO-DATE
:prepareAndroidCordovaLibUnspecifiedReleaseLibrary UP-TO-DATE
:preDebugBuild
:CordovaLib:preDebugBuild UP-TO-DATE
:CordovaLib:compileDebugNdk UP-TO-DATE
:CordovaLib:copyDebugLint UP-TO-DATE
:CordovaLib:mergeDebugProguardFiles UP-TO-DATE
:CordovaLib:packageDebugRenderscript UP-TO-DATE
:CordovaLib:checkDebugManifest
:CordovaLib:prepareDebugDependencies
:CordovaLib:compileDebugRenderscript UP-TO-DATE
:CordovaLib:generateDebugResValues UP-TO-DATE
:CordovaLib:generateDebugResources UP-TO-DATE
:CordovaLib:packageDebugResources UP-TO-DATE
:CordovaLib:compileDebugAidl UP-TO-DATE
:CordovaLib:generateDebugBuildConfig UP-TO-DATE
:CordovaLib:generateDebugAssets UP-TO-DATE
:CordovaLib:mergeDebugAssets UP-TO-DATE
:CordovaLib:processDebugManifest UP-TO-DATE
:CordovaLib:processDebugResources UP-TO-DATE
:CordovaLib:generateDebugSources UP-TO-DATE
:CordovaLib:compileDebugJavaWithJavac UP-TO-DATE
:CordovaLib:processDebugJavaRes UP-TO-DATE
:CordovaLib:transformResourcesWithMergeJavaResForDebug UP-TO-DATE
:CordovaLib:transformClassesAndResourcesWithSyncLibJarsForDebug UP-TO-DATE
:CordovaLib:mergeDebugJniLibFolders UP-TO-DATE
:CordovaLib:transformNative_libsWithMergeJniLibsForDebug UP-TO-DATE
:CordovaLib:transformNative_libsWithSyncJniLibsForDebug UP-TO-DATE
:CordovaLib:bundleDebug UP-TO-DATE
:prepareComAndroidSupportSupportV132340Library UP-TO-DATE
:prepareComAndroidSupportSupportV42400Beta1Library UP-TO-DATE
:prepareReleaseDependencies
:compileReleaseAidl UP-TO-DATE
:compileReleaseRenderscript UP-TO-DATE
:generateReleaseBuildConfig UP-TO-DATE
:generateReleaseAssets UP-TO-DATE
:mergeReleaseAssets
:generateReleaseResValues UP-TO-DATE
:generateReleaseResources UP-TO-DATE
:mergeReleaseResources
AAPT: libpng error: Not a PNG file
AAPT err(Facade for 1201783859) : No Delegate set : lost message:libpng error: Not a PNG file
AAPT err(Facade for 1951788743) : No Delegate set : lost message:libpng error: Not a PNG file
:mergeReleaseResources FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':mergeReleaseResources'.
> Crunching Cruncher icon.png failed, see logs

* Try:
Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output.

BUILD FAILED

```

**Comment by sahanDissanayake** [comment]
@VioletaCalvo what size was the cordova-build-override/platforms/android/res/drawable/pushicon.png Please ?

**Comment by challett** [comment]
@sahanDissanayake 
By the look of the snippet [Here](https://github.com/raix/push/blob/f08079304c1b69897ed05edc7d96954d0416e35e/plugin/push.configuration.js#L100-L106), your icon and iconColor should be at the root of the object, not under gcm.

**Comment by sahanDissanayake** [comment]
ahh!!! Ok let me give that a go in the next couple of days when I get back to get it. Thanks

**Comment by mcoenca** [comment]
@sahanDissanayake @VioletaCalvo Using the tool Android Asset Studio - Notification Icon Generator, i get a cordova-build-override folder that looks like this
![image](https://cloud.githubusercontent.com/assets/6507140/19312052/254fbe2c-9091-11e6-81b2-3e110d2618e5.png)

And with a server config like this,
![image](https://cloud.githubusercontent.com/assets/6507140/19312368/615782aa-9092-11e6-8334-bf073dc3e6f2.png)

I can't get the notif to show the right icon and background...
Any idea ? :(

Android 5.1 and raix:push@3.0.3-rc.7

**Comment by mcoenca** [comment]
Of course I found my stupid mistake 30 seconds after writing my comment ;)
For future readers: all the icons config options should obviously be in CLIENT config. There is no more config.push.json mixing things up like sahanDissanayaka described before in

```
 "gcm": {
    "apiKey": "AIzaSyD5KCLx0q3W-9JXTZI800pYDaatIZDH9a2AY",
    "projectNumber": 107070611301087,
    "icon": "pushicon",
    "iconColor": "#FED121"
  },
```
Now you've got something like
![image](https://cloud.githubusercontent.com/assets/6507140/19312997/2db3c7e4-9095-11e6-9868-4c6196fe01de.png)

**Comment by VioletaCalvo** [comment]
@mcoenca I set up them in the `config.push.json` in the app root folder, you can do that way too. See [here](https://github.com/raix/push/blob/master/docs/ADVANCED.md#setting-credentials--certificates) the full advanced doc.

```json
{
  "apn": {
    "passphrase": "xxxxxxxxxxxxxxxx",
    "key": "APPKey_prod.pem",
    "cert": "APPCert_prod.pem"
  },
  "apn-dev": {
    "passphrase": "xxxxxxxxxxxxxxxx",
    "key": "APPKey_dev.pem",
    "cert": "APPCert_dev.pem"
  },
  "gcm": {
    "apiKey":"xxxxxxxxxxxxxxxxxxxxxxxxxxx",
    "projectNumber":"00000000000"
  },
  "production":false,
  "badge": true,
  "sound": true,
  "alert": true,
  "vibrate": true,
  "icon": "pushiconfilename",
  "iconColor": "#6B97AF"
}
```

@sahanDissanayake sorry, I didn't see your question.
